DM-36517: Add support for ChainedDatastore.export()

timj opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

timj committed:
Add datastore.knows_these() method

This method can run the knows() method in bulk.

timj committed:
Add ChainedDatastore export method

Adjust the test inheritance such that the export tests are
now run for the ChainedDatastore

timj committed:
Add news fragment

timj committed:
Add explicit test for datastore.export

This required we change the exception type in some failure
modes to match that raise by butler.export.

andy-slac commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -262,6 +263,18 @@ def knows(self, ref: DatasetRef) -> bool:
+
+            # No need to check in next datastore for refs that are known.
+            # We only update entries that were initially False.
+            refs = [ref for ref, known in known.items() if not known]
```
`known in known.items() if not known` - I know 😉 that this should work OK, but maybe rename loop variable into something else to reduce potential confusion?

timj replied:
Wow. Not sure what I was thinking with that variable naming...

andy-slac commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -934,3 +947,75 @@ def export_records(self, refs: Iterable[DatasetIdRef]) -> Mapping[str, Datastore
+        refs: Iterable[DatasetRef],
+        *,
+        directory: Optional[ResourcePathExpression] = None,
+        transfer: Optional[str] = "auto",
```
Is it OK to change the default value of `transfer` from the base class?

timj replied:
Why not? I think the base class was written before I came up with the "auto" meaning (when `transfer=None` means "don't transfer" and we determined that was generally a bad default but the base class doesn't really have a choice because we haven't used an enum for the transfer options.

andy-slac replied:
I was thinking that the client can call the method assuming that default for `transfer` is `None`, as declared in the base class, but in reality `transfer` will be "auto" (not sure if there is a difference between the two).

timj replied:
For a chained datastore `transfer=None` could be a bit tricky. `None` means "do not export any files to the given directory, leave them in place" which is fine if you have a single datastore but in a chained datastore you could end up with some entries from one datastore and also entries from another. I suppose I could switch to "auto" if `directory != None` rather than raise below in that scenario.

andy-slac commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -934,3 +947,75 @@ def export_records(self, refs: Iterable[DatasetIdRef]) -> Mapping[str, Datastore
+        missing_count = 0
+        for knows_this in combined_known.values():
+            if not knows_this:
+                missing_count += 1
```
Maybe something like 

    missing_count = list(combined_known.values()).count(False)

going to be shorter/faster?

andy-slac commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -934,3 +947,75 @@ def export_records(self, refs: Iterable[DatasetIdRef]) -> Mapping[str, Datastore
+            for ref, knows_this in known_to_this.items():
+                # If we have not known it yet, store the newer value.
+                if not combined_known.get(ref):
+                    combined_known[ref] = knows_this
```
Maybe use `collections.Counter` for `combined_known`, you can just sum 'True' values with `update()` instead of doing an inner loop (but internally it's probably the same loop).

timj replied:
I'm not really sure I understand how to use `collections.Counter` here. It did make me consider this alternative though:

```python
refs_known: set[DatasetRef] = set()
for known_to_this in known:
    refs_known.update({ref for ref, knows_this in known_to_this.items() if knows_this})
missing_count = len(refs) - len(refs_known)
```
which is much shorter and seems to work.

andy-slac commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -934,3 +947,75 @@ def export_records(self, refs: Iterable[DatasetIdRef]) -> Mapping[str, Datastore
+                exported[ref_positions.pop(ref)] = export
+
+        for i, dataset in enumerate(exported):
+            if dataset is None:
+                raise FileNotFoundError(f"Failed to export dataset {refs[i]}.")
```
Would it make sense to check that all datasets were exported before yielding any of them?

timj replied:
I was trying to avoid scanning through the list of refs an additional time but I can do that.

timj replied:
Thinking about it a bit more, this check is mostly there for mypy so it knows the method can't yield `None`.  The missing check earlier in the method should be enough to know that every dataset is known to at least one datastore. I can add a note that this is for mypy.

andy-slac reviewed (approved):
Looks good, few minor questions.

timj committed:
Simplify logic for noticing missing refs for chained datastore

Also include test with a missing dataset.

timj committed:
Add note that the if check is for mypy

timj committed:
Explicitly default transfer to auto in base class

This requires that auto be converted to None if no directory
is given. FileDatastore and ChainedDatastore already had "auto"
as a default.

timj committed:
Fix yaml type annotation

We had not noticed it was wrong until the pyyaml types package
added an annotation for loader.construct_mapping()
